XMonad/Prelude.hs

```diff
@@ -108,14 +111,19 @@ safeGetWindowAttributes w = withDisplay $ \dpy -> io . alloca $ \p ->
 -- * If it starts with @~\/@, replace that with the actual home
 -- * directory.
 --
--- * Otherwise, prepend a @\/@ to the path.
+-- * If it starts with @$@, read the name of an environment
+-- * variable and replace it with the contents of that.
+--
+-- * Otherwise, prepend the home directory and @\/@ to the path.
 mkAbsolutePath :: MonadIO m => FilePath -> m FilePath
 mkAbsolutePath ps = do
-  home <- liftIO getHomeDirectory
-  pure $ case ps of
-    '/'       : _ ->                ps
-    '~' : '/' : _ -> home <> drop 1 ps
-    _             -> home <> ('/' : ps)
+  home <- io getHomeDirectory
+  case ps of
+    '/'       : _ -> pure ps
+    '~' : '/' : _ -> pure (home <> drop 1 ps)
+    '$'       : _ -> let (v,ps') = span (`elem` ("_"<>['A'..'Z']<>['a'..'z']<>['0'..'9'])) (drop 1 ps)
```

Comment:
I wonder whether we should go through the trouble of expanding all env vars (even not at the start of the string), or if that use-case is too niche

Reply:
Something like (using `X.U.Parser`)

``` haskell
pFilePath :: IO (Parser String)
pFilePath = do
  env <- Map.fromList <$> getEnvironment
  let go :: Parser [String]
      go = many1 $ get >>= \case
        '$' -> pEnvVar
        s   -> (s :) <$> munch1 (/= '$')

      pEnvVar :: Parser String
      pEnvVar = do
        s <- choice [varName, string "{" *> varName <* string "}"]
        maybe empty pure (env Map.!? s)
       where
        varName    = munch1 (`elem` validChars)
        validChars = "_" <> ['1' .. '9'] <> ['A' .. 'Z'] <> ['a' .. 'z']
  pure $ fmap concat go
```

might work, but it might also be overkill.

Reply:
I've been thinking about this (and expanding `~` support to include `~username`), and currently think we can add that later if someone requests it. Adding a full parser for envars, and in particular (possibly repeatedly) sucking the full list into a Map when probably only one will be used, seems massive overkill.

Reply:
Yeah that's fair enough